### PR TITLE
Record unparsable steps in tape

### PR DESCRIPTION
**Pull request by ollmer**
- Update `LLMOutputParsingFailureAction` step to store the full llm output that was the reason of failure.
- Use it during tape reuse to generate corresponding training sample completion
- Small environment module cleanup
- update completion generation for cases with the steps not from llm output

### CODE REVIEW

File: tapeagents/agent.py [hunk 1]
```diff
@@ -392,6 +393,11 @@ def reuse(self, tape: TapeType) -> tuple[TapeType, list[LLMCall]]:
             if self.is_agent_step(step):
                 current_agent = self.delegate(past_tape)
                 prompt = current_agent.make_prompt(past_tape)
+                if not prompt:
+                    logger.debug("Skipping step because agent did not call the llm")
```
Comment (rizar) [inline tapeagents/agent.py:397]: this is a fairly routine event, I don't think it is worth logging

File: tapeagents/nodes.py [hunk 2]
```diff
@@ -52,7 +52,16 @@ def prepare_tape(self, tape: Tape) -> Tape:
         return tape.model_copy(update=dict(steps=steps_without_control_flow))
 
     def make_llm_output(self, agent: Any, tape: Tape, index: int) -> LLMOutput:
-        return LLMOutput(role="assistant", content=tape.steps[index].llm_view())
+        steps = []
```
Comment (rizar) [inline tapeagents/nodes.py:55]: can you add a docstring that explains that what we do here is a heuristic, and how it works?

File: tapeagents/nodes.py [hunk 3]
```diff
@@ -52,7 +52,16 @@ def prepare_tape(self, tape: Tape) -> Tape:
...
+        i = index
+        first_prompt_id = tape.steps[i].metadata.prompt_id
+        while i < len(tape) and tape.steps[i].metadata.prompt_id == first_prompt_id:
+            if not isinstance(tape.steps[i], SetNextNode):
+                steps.append(tape.steps[i])
+            i += 1
+
+        content = [step.llm_dict() for step in steps] if len(steps) > 1 else steps[0].llm_dict()
```
Comment (rizar) [inline tapeagents/nodes.py:63]: I think we need `steps[-1].llm_dict()` here instead

**Review by rizar — CHANGES_REQUESTED** [review]
not sure if the code is correct, let's discuss

File: tapeagents/nodes.py (hunk 3 above)
Comment (ollmer) [reply tapeagents/nodes.py:63]: Some of our agents produce more than 1 step per call. It is correctly handled by tracking the same prompt_id.

**Pull request merged**